cudaextender/src/ungapped_xdrop_kernels.cu

```diff
@@ -0,0 +1,459 @@
+            {
+                new_max_found[warp_id] = max_pos > prev_max_pos[warp_id];
+                prev_max_pos[warp_id]  = max_pos;
+                if (xdrop_done)
+                {
+                    total_score[warp_id] += max_thread_score;
+                    xdrop_found[warp_id] = true;
+                    left_extent[warp_id] = max_pos;
+                    extent[warp_id] += left_extent[warp_id];
+                    tile[warp_id] = max_pos;
+                }
+                else if (ref_loc[warp_id] < pos_offset || query_loc[warp_id] < pos_offset)
+                {
+                    total_score[warp_id] += max_thread_score;
+                    edge_found[warp_id]  = true;
+                    left_extent[warp_id] = max_pos;
+                    extent[warp_id] += left_extent[warp_id];
+                    tile[warp_id] = max_pos;
+                }
+                else
+                {
+                    prev_score[warp_id]     = thread_score;
+                    prev_max_score[warp_id] = max_thread_score;
+                    tile[warp_id] += warpSize;
+                }
+            }
+            __syncwarp();
+
+            if (new_max_found[warp_id])
+            {
+#pragma unroll
+                for (int32_t i = 0; i < 4; i++)
+                {
+                    count[i]     = count[i] + count_del[i];
+                    count_del[i] = 0;
+                }
+            }
+            __syncwarp();
+
+            if (r_chr == q_chr)
+            {
+                if (pos_offset <= prev_max_pos[warp_id])
+                {
+                    count[r_chr] += 1;
+                }
+                else
+                {
+                    count_del[r_chr] += 1;
+                }
+            }
+            __syncwarp();
+        }
+
+        //////////////////////////////////////////////////////////////////
+
+        if (total_score[warp_id] >= score_threshold && total_score[warp_id] <= 3 * score_threshold && !no_entropy)
+        {
+            for (int32_t i = 0; i < 4; i++)
+            {
+#pragma unroll
+                for (int32_t offset = 1; offset < warpSize; offset = offset << 1)
+                {
+                    count[i] += __shfl_up_sync(0xFFFFFFFF, count[i], offset);
+                }
+            }
+            __syncwarp();
+
+            if (lane_id == warpSize - 1 && ((count[0] + count[1] + count[2] + count[3]) >= 20)) // TODO - MAGIC NUMBER ALERT!
+            {
+
+                entropy[warp_id] = 0.f;
+#pragma unroll
+                for (int32_t i = 0; i < 4; i++)
+                {
+                    if (count[i] != 0)
+                    {
+                        const double probability = static_cast<double>(count[i]) / static_cast<double>(extent[warp_id] + 1);
+                        entropy[warp_id] += (probability)*log(probability);
+                    }
+                }
+                entropy[warp_id] = -entropy[warp_id] / log_4;
```

Comment:
I've got a hunch that here might be a bug. What's the mathematical formula for the entroy?

Reply:
The traditional formula is:
 `-sum(probability.*log2(probability))`  
The log base depends on units we're trying to measure the entropy for. 2 is for bits, and I'm guessing 4 here is for the ACGT alphabet.

Reply:
Right, never mind...